src/NHibernate/Proxy/ProxyBuilderHelper.cs

```diff
@@ -134,25 +136,48 @@ internal static MethodBuilder GetObjectDataMethodBuilder(TypeBuilder typeBuilder
 			return methodBuilder;
 		}
 
+		private static readonly System.Type[] _equalsParametersTypes = { typeof(object) };
+
 		internal static MethodBuilder GenerateMethodSignature(string name, MethodInfo method, TypeBuilder typeBuilder)
 		{
-			//TODO: Should we use attributes of base method?
-			var methodAttributes = MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.Virtual;
+			var explicitImplementation = method.DeclaringType.IsInterface;
+			if (explicitImplementation &&
+				(typeBuilder.BaseType == typeof(object) ||
+#pragma warning disable 618
+					typeBuilder.BaseType == typeof(ProxyDummy)) &&
+#pragma warning restore 618
+				(method.Name == "Equals" && method.GetParameters().Select(p => p.ParameterType).SequenceEqual(_equalsParametersTypes) ||
+					method.Name == "GetHashCode" && method.GetParameters().Length == 0))
+			{
+				// If we are building a proxy for an interface, and it defines an Equals or GetHashCode, they must
+				// be implicitly implemented for overriding object methods.
+				// (Ideally we should check the method actually comes from the interface declared for the proxy.)
+				explicitImplementation = false;
+			}
+
+			var methodAttributes = explicitImplementation
+				? MethodAttributes.Private | MethodAttributes.Final | MethodAttributes.HideBySig |
+					MethodAttributes.SpecialName | MethodAttributes.NewSlot | MethodAttributes.Virtual
+				: MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.Virtual;
 
 			if (method.IsSpecialName)
 				methodAttributes |= MethodAttributes.SpecialName;
 
 			var parameters = method.GetParameters();
-
-			var methodBuilder = typeBuilder.DefineMethod(
-				name,
-				methodAttributes,
-				CallingConventions.HasThis,
-				method.ReturnType,
-				parameters.ToArray(param => param.ParameterType));
+			var implementationName = explicitImplementation
+				? $"{method.DeclaringType.FullName}.{name}"
+				: name;
+			var methodBuilder =
+				typeBuilder.DefineMethod(
+					implementationName,
+					methodAttributes,
+					CallingConventions.HasThis,
+					method.ReturnType,
+					parameters.ToArray(param => param.ParameterType));
+			if (explicitImplementation)
+				methodBuilder.SetImplementationFlags(MethodImplAttributes.Managed | MethodImplAttributes.IL);
```

Comment:
Copied from `NHibernateProxyBuilder.ImplementGetLazyInitializer` explicit implementation, I do not actually know if it is needed.